File: src/main/java/com/freya02/bot/commands/slash/docs/DocsCommand.java
package com.freya02.bot.commands.slash.docs;

import com.freya02.bot.docs.DocIndexMap;
import com.freya02.bot.docs.index.DocIndex;
import com.freya02.botcommands.api.BContext;
import com.freya02.botcommands.api.application.ApplicationCommand;
import com.freya02.botcommands.api.application.CommandPath;
import com.freya02.botcommands.api.application.annotations.AppOption;
import com.freya02.botcommands.api.application.slash.DefaultValueSupplier;
import com.freya02.botcommands.api.application.slash.GuildSlashEvent;
import com.freya02.botcommands.api.application.slash.annotations.JDASlashCommand;
import com.freya02.docs.DocSourceType;
import net.dv8tion.jda.api.entities.Guild;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;

public class DocsCommand extends ApplicationCommand {
	private final DocIndexMap docIndexMap;

	public DocsCommand() throws IOException {
		docIndexMap = DocIndexMap.getInstance();
	}

	@Override
	@Nullable
	public DefaultValueSupplier getDefaultValueSupplier(@NotNull BContext context, @NotNull Guild guild,
	                                                    @Nullable String commandId, @NotNull CommandPath commandPath,
	                                                    @NotNull String optionName, @NotNull Class<?> parameterType) {
		if (optionName.equals("source_type")) {
			//use subcommand as a default value
			if ("botcommands".equals(commandPath.getSubname())) {
				return e -> DocSourceType.BOT_COMMANDS;
			} else if ("jda".equals(commandPath.getSubname())) {
				return e -> DocSourceType.JDA;
			} else if ("java".equals(commandPath.getSubname())) {
				return e -> DocSourceType.JAVA;
			}
		}

		return super.getDefaultValueSupplier(context, guild, commandId, commandPath, optionName, parameterType);
	}

	@JDASlashCommand(
			name = "docs",
			subcommand = "botcommands",
			description = "Shows the documentation for a class, a method or a field"
	)
	public void onSlashDocsBC(@NotNull GuildSlashEvent event,
	                          @NotNull @AppOption(description = "The docs to search upon")
			                          DocSourceType sourceType,
	                          @NotNull @AppOption(description = "Name of the Java class", autocomplete = CommonDocsHandlers.CLASS_NAME_WITH_METHODS_AUTOCOMPLETE_NAME)
			                          String className,
	                          @Nullable @AppOption(description = "Signature of the method / field name", autocomplete = CommonDocsHandlers.METHOD_NAME_OR_FIELD_BY_CLASS_AUTOCOMPLETE_NAME)
			                          String identifier) throws IOException {
		onSlashDocs(event, sourceType, className, identifier);
	}

	@JDASlashCommand(
			name = "docs",
			subcommand = "jda",
			description = "Shows the documentation for a class, a method or a field"
	)
	public void onSlashDocsJDA(@NotNull GuildSlashEvent event,
	                           @NotNull @AppOption(description = "The docs to search upon")
			                           DocSourceType sourceType,
	                           @NotNull @AppOption(description = "Name of the Java class", autocomplete = CommonDocsHandlers.CLASS_NAME_WITH_METHODS_AUTOCOMPLETE_NAME)
			                           String className,
	                           @Nullable @AppOption(description = "Signature of the method / field name", autocomplete = CommonDocsHandlers.METHOD_NAME_OR_FIELD_BY_CLASS_AUTOCOMPLETE_NAME)
			                           String identifier) throws IOException {
		onSlashDocs(event, sourceType, className, identifier);
	}

	@JDASlashCommand(
			name = "docs",
			subcommand = "java",
			description = "Shows the documentation for a class, a method or a field"
	)
	public void onSlashDocsJava(@NotNull GuildSlashEvent event,
	                            @NotNull @AppOption(description = "The docs to search upon")
			                            DocSourceType sourceType,
	                            @NotNull @AppOption(description = "Name of the Java class", autocomplete = CommonDocsHandlers.CLASS_NAME_WITH_METHODS_AUTOCOMPLETE_NAME)
			                            String className,
	                            @Nullable @AppOption(description = "Signature of the method / field name", autocomplete = CommonDocsHandlers.METHOD_NAME_OR_FIELD_BY_CLASS_AUTOCOMPLETE_NAME)
			                            String identifier) throws IOException {
		onSlashDocs(event, sourceType, className, identifier);
	}

	private void onSlashDocs(@NotNull GuildSlashEvent event,
	                         @NotNull DocSourceType sourceType,
	                         @NotNull String className,
	                         @Nullable String identifier) throws IOException {

		final DocIndex docIndex = docIndexMap.get(sourceType);

		if (identifier == null) {
			CommonDocsHandlers.handleClass(event, className, docIndex);
		} else if (identifier.contains("(")) { //prob a method
			CommonDocsHandlers.handleMethodDocs(event, className, identifier, docIndex);
		} else {
			CommonDocsHandlers.handleFieldDocs(event, className, identifier, docIndex);
		}
	}
}

File: src/main/kotlin/com/freya02/bot/commands/slash/docs/DocsCommand.kt
package com.freya02.bot.commands.slash.docs

import com.freya02.bot.commands.slash.docs.CommonDocsHandlers.Companion.CLASS_NAME_AUTOCOMPLETE_NAME
import com.freya02.bot.commands.slash.docs.CommonDocsHandlers.Companion.EXPERIMENTAL_SEARCH_AUTOCOMPLETE_NAME
import com.freya02.bot.commands.slash.docs.CommonDocsHandlers.Companion.METHOD_OR_FIELD_BY_CLASS_AUTOCOMPLETE_NAME
import com.freya02.bot.commands.slash.docs.CommonDocsHandlers.Companion.SEARCH_AUTOCOMPLETE_NAME
import com.freya02.bot.commands.slash.docs.controllers.SlashDocsController
import com.freya02.bot.docs.DocIndexMap
import com.freya02.bot.docs.index.DocSuggestion
import com.freya02.bot.docs.index.DocSuggestion.Companion.mapToSuggestions
import com.freya02.bot.docs.index.DocTypes
import com.freya02.botcommands.api.annotations.CommandMarker
import com.freya02.botcommands.api.commands.application.CommandScope
import com.freya02.botcommands.api.commands.application.GuildApplicationCommandManager
import com.freya02.botcommands.api.commands.application.annotations.AppDeclaration
import com.freya02.botcommands.api.commands.application.slash.GuildSlashEvent
import com.freya02.docs.DocSourceType

@CommandMarker
class DocsCommand(private val docIndexMap: DocIndexMap, private val slashDocsController: SlashDocsController) {
    @AppDeclaration
    fun declare(manager: GuildApplicationCommandManager) {
        manager.slashCommand("docs", CommandScope.GUILD) {
            description = "Shows the documentation"

            DocSourceType.typesForGuild(manager.guild).forEach { sourceType ->
                subcommand(sourceType.cmdName) {
                    description = "Shows the documentation for a class, a method or a field"

                    generatedOption("sourceType") { sourceType }

                    option("className") {
                        description = "Name of the Java class"
                        autocompleteReference(CLASS_NAME_AUTOCOMPLETE_NAME)
                    }

                    option("identifier") {
                        description = "Signature of the method / field name"
                        autocompleteReference(METHOD_OR_FIELD_BY_CLASS_AUTOCOMPLETE_NAME)
                    }

                    function = ::onSlashDocs
                }
            }

            subcommandGroup("search") {
                DocSourceType.typesForGuild(manager.guild).forEach { sourceType ->
                    subcommand(sourceType.cmdName) {
                        description = "Searches the documentation for anything"

                        generatedOption("sourceType") { sourceType }

                        //Required for autocomplete
                        generatedOption(declaredName = "docTypes") { DocTypes.ANY }

                        option("query") {
                            description = "The docs to search for"
                            autocompleteReference(SEARCH_AUTOCOMPLETE_NAME)
                        }

                        function = slashDocsController::onSearchSlashCommand
                    }
                }
            }

            subcommandGroup("exp") {
                DocSourceType.typesForGuild(manager.guild).forEach { sourceType ->
                    subcommand(sourceType.cmdName) {
                        description = "Experimental - Searches the documentation for anything, with a custom sort"

                        generatedOption("sourceType") { sourceType }

                        //Required for autocomplete
                        generatedOption(declaredName = "docTypes") { DocTypes.ANY }

                        option("query") {
                            description = "The docs to search for"
                            autocompleteReference(EXPERIMENTAL_SEARCH_AUTOCOMPLETE_NAME)
                        }

                        function = slashDocsController::onSearchSlashCommand
                    }
                }
            }
        }
    }

    @CommandMarker
    suspend fun onSlashDocs(
        event: GuildSlashEvent,
        sourceType: DocSourceType,
        className: String,
        identifier: String?
    ) {
        val docIndex = docIndexMap[sourceType]!!
        if (identifier == null) {
            slashDocsController.handleClass(event, className, docIndex) {
                return@handleClass classNameAutocomplete(docIndex, className, 100)
                    .map { DocSuggestion(it, it) }
            }
        } else if (identifier.contains("(")) { //prob a method
            slashDocsController.handleMethodDocs(event, className, identifier, docIndex) {
                return@handleMethodDocs methodOrFieldByClassAutocomplete(docIndex, className, identifier, 100).mapToSuggestions()
            }
        } else {
            slashDocsController.handleFieldDocs(event, className, identifier, docIndex) {
                return@handleFieldDocs methodOrFieldByClassAutocomplete(docIndex, className, identifier, 100).mapToSuggestions()
            }
        }
    }
}
